libraries/state_history/include/eosio/state_history/log_catalog.hpp

```diff
@@ -0,0 +1,261 @@
+   struct by_mru {};
+   typedef multi_index_container<
+     catalogued_log_file,
+     indexed_by<
+       ordered_unique<key<&catalogued_log_file::begin_block_num>>,
+       ordered_non_unique<tag<by_mru>,key<&catalogued_log_file::effective_last_used_counter>, std::greater<size_t>>
+     >
+   > catalog_t;
+   catalog_t retained_log_files;
+   std::optional<state_history_log> head_log;
+   const std::filesystem::path head_log_path_and_basename;  //example: /some/dir/trace-history   i.e. does NOT include .log nor .index
+
+   size_t global_used_counter = 0;
+
+public:
+   log_catalog(const log_catalog&) = delete;
+   log_catalog& operator=(log_catalog&) = delete;
+
+   log_catalog(const std::filesystem::path& log_dir, const state_history::state_history_log_config& config, const std::string& log_name,
+               state_history_log::non_local_get_block_id_func non_local_get_block_id = state_history_log::no_non_local_get_block_id_func) :
+     non_local_get_block_id(non_local_get_block_id), head_log_path_and_basename(log_dir / log_name) {
+      std::visit(chain::overloaded {
+         [this](const std::monostate&) {
+            open_head_log();
+         }, //nothing needed
+         [this](const state_history::prune_config& prune) {
+            open_head_log(prune);
+         },
+         [this, &log_dir, &log_name](const state_history::partition_config& partition_config) {
+            retained_dir = make_absolute_dir(log_dir, partition_config.retained_dir.empty() ? log_dir : partition_config.retained_dir);
+            if(!partition_config.archive_dir.empty())
+               archive_dir = make_absolute_dir(log_dir, partition_config.archive_dir);
+            max_retained_files = partition_config.max_retained_files;
+            log_rotation_stride = partition_config.stride;
+
+            const std::regex retained_logfile_regex("^" + log_name + R"(-\d+-\d+\.log$)");
+
+            for(const std::filesystem::directory_entry& dir_entry : std::filesystem::directory_iterator(retained_dir)) {
+               if(!dir_entry.is_regular_file())
+                  continue;
+               if(!std::regex_search(dir_entry.path().filename().string(), retained_logfile_regex))
+                  continue;
+
+               const std::filesystem::path path_and_basename = dir_entry.path().parent_path() / dir_entry.path().stem();
+
+               state_history_log log(path_and_basename, [](chain::block_num_type) {return std::nullopt;});
+               if(log.empty())
+                  continue;
+               const auto [begin_bnum, end_bnum] = log.block_range();
+               retained_log_files.emplace(begin_bnum, end_bnum, path_and_basename);
+            }
+
+            if(retained_log_files.size() > 1)
+               for(decltype(retained_log_files)::iterator it = retained_log_files.begin(); it != std::prev(retained_log_files.end()); ++it)
```

Comment:
It feels like the definition of `retained_log_files` should make that fairly clear,
https://github.com/AntelopeIO/spring/blob/a29f3f96aa9bfebc3a33eb4c3578bab74bb08527/libraries/state_history/include/eosio/state_history/log_catalog.hpp#L51-L59
specifically `ordered_unique<key<&catalogued_log_file::begin_block_num>>` 

Something we could do is attach a tag to this index like `by_begin_block` so that everywhere we use multiindex we explicitly call out `by_begin_block`. Without indirection from more `typedef`s it'd look something like
```c++
for(catalog_t::index<by_begin_block>::type::iterator it = retained_log_files.get<by_begin_block>().begin(); it != std::prev(retained_log_files.get<by_begin_block>().end()); ++it)
```
But this feels awfully wordy to me, with what is already quite a mouthful as is